### ISSUE TITLE
Fase 1 (FD001) - Feature Engineering: Ventanas temporales y selección de características

### ISSUE THREAD

**Issue by JairRamosG** [JairRamosG/RUL-CMAPSS#5]
## Goal

Implementar el módulo profesional y desacoplado de **Selección de Características** (`src/features/selection.py`) bajo un enfoque híbrido en dos etapas (**Mutual Information** como filtro univariado rápido + **Random Forest Feature Importance** como selector embebido multivariado), integrándolo dentro del pipeline de evaluación (`scripts/train_eval.py`) respetando estrictamente el protocolo anti-leakage dentro de cada fold de validación cruzada (`GroupKFold`).

---

## Pipeline Flow & Protocolo Anti-Leakage (Design Decision)

Siguiendo las mejores prácticas de la industria y la literatura estadística (Hastie, Tibshirani & Friedman, 2009, *ESL* §7.10.2):
* **Cualquier selección supervisada que consulte la variable objetivo ($y$) DEBE ejecutarse exclusivamente dentro del conjunto de entrenamiento de cada partición (`X_train_fold`).**
* Las features resultantes de cada fold son compartidas idénticamente por todos los modelos evaluados, garantizando la validez del **diseño experimental de bloques pareados** (Friedman / RM-ANOVA).

```
Raw Data (100 motores)
       │
       ▼
Feature Engineering (Sensores + Rolling Stats + Trends -> 102 columnas)
       │
       ▼
GroupKFold Split (k=10 por unit_number)
       │
       ├── Para cada Fold i (Train: 90 motores | Val: 10 motores):
       │     ├── 1. Escalado: MinMaxScaler ajustado solo en train_fold
       │     ├── 2. Feature Selection Pipeline (fit en train_fold):
       │     │        ├── Etapa 1 (Filtro): Mutual Information (elimina ruido / baja información)
       │     │        └── Etapa 2 (Embebido): Random Forest MDI (selecciona mejores k features con interacciones)
       │     ├── 3. Transformación: Selección aplicada a train_fold y val_fold
       │     ├── 4. create_windows(): Generación de tensores 3D (N, W=30, F_sel)
       │     └── 5. Entrenamiento y evaluación pareada de los 7 modelos sobre F_sel
       │
       ▼
Evaluación Final en Test Set Oficial:
       ├── Pipeline.fit() sobre el 100% de train (100 motores) -> F_sel_global
       └── Re-entrenamiento final y predicción sobre último ciclo de test
```

---

### 1. Transformadores Atómicos Desacoplados (`src/features/selection.py`)
- [ ] **`MutualInfoSelector(BaseEstimator, TransformerMixin)`**:
  - Filtro univariado no lineal basado en `sklearn.feature_selection.mutual_info_regression`.
  - Parámetros: `percentile` (int, default 60) o `threshold` (float, default None).
  - Semilla determinista: `random_state=42`.
  - Métodos estándar: `fit(X, y)`, `transform(X)`, `fit_transform(X, y)`, `get_support()`, `get_feature_names_out()`.
- [ ] **`RFFeatureSelector(BaseEstimator, TransformerMixin)`**:
  - Selector embebido multivariado basado en `RandomForestRegressor`.
  - Parámetros: `n_features_to_select` (int, default 25), `n_estimators` (int, default 100), `max_depth` (default 15), `random_state=42`, `n_jobs=-1`.
  - Selección basada en `feature_importances_` (MDI).
  - Métodos estándar: `fit(X, y)`, `transform(X)`, `fit_transform(X, y)`, `get_support()`, `get_feature_names_out()`.

### 2. Composición y Factory Híbrida (`src/features/selection.py`)
- [ ] Función `create_feature_selector(method="hybrid", config=None) -> Pipeline | BaseEstimator`:
  - Permite configurar `"mutual_info"`, `"rf_importance"` o `"hybrid"`.
  - Si es `"hybrid"`, retorna un `sklearn.pipeline.Pipeline([("filter", MutualInfoSelector(...)), ("embedded", RFFeatureSelector(...))])`.
  - Si es `None` o método desactivado, retorna un `FunctionTransformer` (passthrough sin alterar columnas).

### 3. Configuración en YAML (`configs/config_FD001.yaml`)
- [ ] Nueva sección configurable:
  ```yaml
  feature_selection:
    enabled: true
    method: "hybrid" # "mutual_info" | "rf_importance" | "hybrid"
    mutual_info:
      percentile: 60
    rf_importance:
      n_features_to_select: 25
      n_estimators: 100
      random_state: 42
  ```

### 4. Integración en el Pipeline Principal (`scripts/train_eval.py`)
- [ ] Inyección de `feature_selector.fit(X_train_df, y_train)` **dentro de cada fold** tras el escalado y previo al ventaneo temporal `create_windows()`.
- [ ] Ajuste dinámico de `input_shape = (window_size, len(selected_features))` para que tanto modelos clásicos 2D como Deep Learning 3D consuman directamente el tensor reducido `(N, W, F_sel)`.
- [ ] Garantizar que todos los modelos evaluados en el fold $k$ utilicen exactamente las mismas características seleccionadas para dicho fold.
- [ ] Evaluación de Test Set: ajuste del selector con todo el conjunto de entrenamiento y transformación consistente de la partición de test.

### 5. Análisis de Estabilidad de Características (Soporte Tesis)
- [ ] Seguimiento y logueo de las columnas seleccionadas en cada fold.
- [ ] Cálculo de la **frecuencia de selección** de cada feature a través de los 10 folds (ej. features universales presentes en 10/10 folds vs features esporádicas).
- [ ] Registro de la lista de features consolidadas y su frecuencia en el log/run de MLflow.

### 6. Pruebas Unitarias Exhaustivas (`tests/test_selection.py`)
- [ ] Herencia y cumplimiento del contrato de `BaseEstimator` y `TransformerMixin` de Scikit-learn.
- [ ] Verificación de que `fit` falla con error informativo si no se suministra `y` en métodos supervisados.
- [ ] Verificación de formas de tensores y arrays antes y después de `transform`.
- [ ] Verificación de interoperabilidad con `Pipeline` de scikit-learn.
- [ ] Test de invariabilidad de features en test/val (solo proyecta columnas aprendidas en train).

---

## Ya Implementado Previamente (Fase Feature Engineering)
- [x] `create_windows()` en `src/features/engineering.py` (con edge padding).
- [x] `compute_rolling_stats()` en `src/features/engineering.py` (mean, std, min, max).
- [x] `compute_trends()` en `src/features/engineering.py` (deltas de primer orden).
- [x] Tests de ingeniería en `tests/test_engineering.py`.

---

## Constraints & Environment
- Entorno: Linux Mint | Python 3.11+ | CPU-only (`n_jobs=-1` para RF).
- Scikit-learn Pipeline como estándar de composición.
- Cero tolerancia a Data Leakage (fit exclusivo en particiones train).
